### PR TITLE
Fix Bert serialization

### LINKED ISSUE #312: Register our BERT components as a serializable Keras objects
We would like to be able to annotate our new BERT models and components with `@keras.utils.register_keras_serializable(package="keras_nlp")`. This will allow the actual python object to restore after saving, which will be import to access things like `model.tokenizer` or `model.token_embedding` after restoring.

### PR DESCRIPTION
**Pull request by mattdangerw**
Fixes #312

The following changes do a few things, following the advice [here](https://keras.io/guides/serialization_and_saving/#how-savedmodel-handles-custom-objects).

 - Allows us to actually restore the real python Bert object when saving and restoring via tf or h5 formats.
 - Simplify our `get_config()` results so they are much more readable for Bert models, and reflect the actual configurable parts of the symbols.

### CODE REVIEW

File: keras_nlp/models/bert/bert_models.py [hunk 1]
```diff
@@ -248,21 +249,22 @@ def __init__(
...
+        return {
+            "vocabulary_size": self.vocabulary_size,
+            "hidden_dim": self.hidden_dim,
+            "intermediate_dim": self.intermediate_dim,
+            "num_layers": self.num_layers,
+            "num_heads": self.num_heads,
+            "max_sequence_length": self.max_sequence_length,
+            "num_segments": self.num_segments,
+            "dropout": self.dropout,
+            "name": self.name,
```
Comment (fchollet) [inline keras_nlp/models/bert/bert_models.py:261]: You might want to add `self.dtype`

**Review by fchollet — APPROVED** [review]
LGTM

File: keras_nlp/models/bert/bert_models.py [hunk 2]
```diff
@@ -248,21 +249,22 @@ def __init__(
...
+            "trainable": self.trainable,
```
Comment (jbischof) [inline keras_nlp/models/bert/bert_models.py:262]: Why is `trainable` part of the config? Presumably that is something dynamic to adjusted to the context? Not 100% sure about `name` either.

File: keras_nlp/models/bert/bert_tasks.py [hunk 3]
```diff
@@ -104,6 +105,20 @@ def __init__(
...
+            "backbone": keras.layers.serialize(self.backbone),
+            "num_classes": self.num_classes,
+            "name": self.name,
+            "trainable": self.trainable,
+        }
+
+    @classmethod
+    def from_config(cls, config):
+        if "backbone" in config:
+            config["backbone"] = keras.layers.deserialize(config["backbone"])
```
Comment (jbischof) [inline keras_nlp/models/bert/bert_tasks.py:119]: How does `serialize` compare to `get_config`? What would it look like to store a `preset` for a pretrained classifier model in this system? I was imagining an in-lined `config` for the `backbone` but I'm not sure what this serialized format would look like.

**Review by jbischof — COMMENTED** [review]
Looks good, but some small questions.

File: keras_nlp/models/bert/bert_models.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/models/bert/bert_models.py:262]: `get_config` will be used during model saving to propagate those options. You want `name` and `trainable` in there (try this our on a dense layer, for example).

Normally, for a layer, if you chain to `super()` you get all of this stuff. But in our case, we don't want the big nested generic nested config that a functional model gives you by default, because that will give you back a generic functional object, with none of our `Bert` python properties and methods on it.

Ideally, we would fix this all in core Keras, so there was a better supported flow for a "functional subclass." Possibly this would include a way to chain to super before defining your functional layers in `init`, and a way you could simply call `super()` here in `get_config` and get the same behavior as a normal layer. But I don't think we want to gate this on a core Keras feature.

I think without core Keras changes, the ways to do this are to either:
 - Include `name` and `trainable` explicitly.
 - Explicitly call `keras.Layer.get_config(self)` at the top of this function.

File: keras_nlp/models/bert/bert_tasks.py (hunk 3 above)
Comment (mattdangerw) [reply keras_nlp/models/bert/bert_tasks.py:119]: Serialize will include the symbol name in a serialization friendly manner (and I think that's pretty much it?). But in short I think that's it is the more idiomatic way to save a config for nested objects like this.

Here's some examples with model and layers:
https://github.com/keras-team/keras/blob/c269e3cd8fed713fb54d2971319df0bfe6e1bf10/keras/models/sharpness_aware_minimization.py#L147
https://github.com/keras-team/keras/blob/c269e3cd8fed713fb54d2971319df0bfe6e1bf10/keras/premade_models/wide_deep.py#L214
https://github.com/keras-team/keras/blob/c269e3cd8fed713fb54d2971319df0bfe6e1bf10/keras/layers/rnn/stacked_rnn_cells.py#L204

We also do the same for serializing our activations and initializers in a config (everything hits `serialize_keras_object` eventually).

Comment (jbischof) [reply keras_nlp/models/bert/bert_tasks.py:119]: See #377 for some of the nested configs I was imagining. For example, how would you propose we do the following?

```python
classifier_checkpoints = {
    "toxic_bert": {
        "config": {
            "backbone_config": {
                "vocabulary_size": 10000, 
                "hidden_dim": 768, 
                "intermediate_dim": 3072, 
                "num_layers": 12, 
                "num_heads": 12, 
                "max_sequence_length": 512, 
                "num_segments": 2, 
                "dropout": 0.1
        },
            "num_classes": 6,
        },
        "description": (
            "A multi-headed model capable of detecting different types "
            "of toxicity like threats, obscenity, insults, and identity-based hate."
        ),
        "weights_url": "https://storage.googleapis.com/keras-nlp/models/toxic_bert.h5",
        "weights_hash": "9b2b2139f221988759ac9cdd17050b31",
    },
}
```

Comment (mattdangerw) [reply keras_nlp/models/bert/bert_tasks.py:119]: I think it's the same almost! Here's how you would do that with the code in this PR.

```
classifiers = {
    "toxic_bert": {
        "config": {
            "backbone": {
                "class_name": "keras_nlp>Bert",
                "config": {
                    "vocabulary_size": 10000,
                    "num_layers": 24,
                    "num_heads": 16,
                    "hidden_dim": 768,
                    "intermediate_dim": 3072,
                    "dropout": 0.1,
                    "max_sequence_length": 512,
                },
            },
            "num_classes": 2,
        },
    },
}
```

The advantage of this of this we have pushed things onto a standard Keras mechanism. Let's say we add an initializer argument to this and core `Bert` in the future. Like of our layers, we would serialize this with `keras.initalizers.serialize`, and it would show up in the config like this:

```
'backbone': {...},
'initializer': {'class_name': 'GlorotNormal', 'config': {'seed': None}},
'num_classes': 2,
```
or
```
'backbone': {...},
'initializer': {'class_name': 'keras_nlp>BertInitializer', 'config': {...}},
'num_classes': 2,
```

We have made our mechanism for nesting Keras objects uniform with the rest of Keras. And we leave room for other symbol support in the future (maybe not a big deal for BERT in this specific case, but nice to have generally).

Note that if we are really opposed to the serialization extra bits we could choose to avoid this specifically for the `from_preset()` constructor. But for the `from_config()` constructor I think it makes sense to follow normal Keras idioms for sub object serialization.

Comment (jbischof) [reply keras_nlp/models/bert/bert_tasks.py:119]: Thank you for the detailed explanation @mattdangerw!

File: keras_nlp/models/bert/bert_models.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/models/bert/bert_models.py:262]: Ah got you. Yeah they are persisted.
